### PR TITLE
WIP Add spec for libvirt driver launching AMD SEV-encrypted instances

### PR DESCRIPTION
**Pull request by aspiers**
This spec proposes a minimal set of work required in order for nova's libvirt driver to support launching of KVM instances which are encrypted using AMD's SEV (Secure Encrypted Virtualization) technology:

- https://developer.amd.com/sev/

This is just a dummy PR so that SUSE folks can give an initial review before I submit to Gerrit.

### CODE REVIEW

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+
+Problem description
+===================
+
+Cloud customers are greatly concerned about Spectre, Meltdown, and
+many similar vulnerabilities which enable a malicious VM to harvest
+data from another VM running on the same hardware without being
+detected.  What was once a theoretical attack has now been realized
+with the result that customers are concerned that their data may not
+be adequately protected in a cloud environment.  AMD's SEV offers a
```
Comment (bryanstephenson) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: Need to change from "AMD's SEV offers a potential solution for this which"    to    "AMD's SEV offers a new VM protection technology which".

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+it should probably not be relied upon as the sole line of defense.
+
+Use Cases
+---------
+
+#. Booting an instance with a flavor containing an SEV-specific
+   `required trait extra spec
+   <https://docs.openstack.org/nova/latest/user/flavors.html#extra-specs-required-traits>`_
+   will cause that instance to be defined and booted on a compute host
+   which provides that trait.  Additionally the instance should be
```
Comment (bryanstephenson) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: Do we want to change "should" to "must" here?

**Review by bryanstephenson — COMMENTED** [review]
Looks good so far. Just a couple comments.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+References
+==========
+
+Still TODO
+
+Please add any useful references here. You are not required to have any
+reference. Moreover, this specification should still make sense when your
+references are unavailable. Examples of what you could include are:
+
+* Links to mailing list or IRC discussions
```
Comment (bryanstephenson) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: We could add this reference:

* `Slides describing SEV and preliminary performance results <http://events17.linuxfoundation.org/sites/events/files/slides/AMD%20SEV-ES.pdf>`_

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+The end user will harness SEV through the existing mechanisms of
+traits in flavor extra specs and properties.  Later on it may make
+sense to add support for scheduler hints (see the `Future Work`_
+section below).
+
+Performance Impact
+------------------
+
+Encrypted VMs will inevitably perform worse than unencrypted ones
+due to the encryption overhead.
```
Comment (bryanstephenson) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: we could add another sentence:

Preliminary testing indicates that the expected performance impact is moderate; a degradation of 1% to 6% has been observed depending on the particular workload and test (see slides in the Reference section).

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+  document that fact.
+
+* Does this feature require any new library dependencies or code otherwise not
+  included in OpenStack? Or does it depend on a specific version of library?
+
+
+Testing
+=======
+
+Still TODO
```
Comment (bryanstephenson) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: We could involve Jenny Wei on the testing part of this blueprint, for writing and/or review. She has developed additional Nova tests.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+
+====================================================
+libvirt driver launching AMD SEV-encrypted instances
+====================================================
+
+https://blueprints.launchpad.net/nova/+spec/amd-sev
+
+This spec proposes a minimal set of work required in order for nova's
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: s,a minimal set of,, 

from an upstream perspective we shouldn't argue what the "minimal work" is but what the "right amount of implementation is. minimalism vs keeping it simple.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+libvirt driver to support launching of KVM instances which are
+encrypted using `AMD's SEV (Secure Encrypted Virtualization)
+technology <https://developer.amd.com/sev/>`_.
+
+
+Problem description
+===================
+
+Cloud customers are greatly concerned about Spectre, Meltdown, and
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: Again, I don't think its proven that SEV protects from cache sidechannel attacks, because the encryption/decryption is not bypassing caching. as such this statement  is unnecessary and potentially wrong. we don't need to do unnecessary marketing against competing vendors like AMD or intel (some of which claim potentially more complete competing technologies).

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 1 above)
Comment (dirkmueller) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: is it really new? :-)

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+detected.  What was once a theoretical attack has now been realized
+with the result that customers are concerned that their data may not
+be adequately protected in a cloud environment.  AMD's SEV offers a
+potential solution for this which transparently encrypts the memory of
+each VM with a unique key.  It can also calculate a signature of the
+memory contents, which can be sent to the guest owner as an
+attestation that the memory was encrypted correctly by the firmware.
+Therefore the solution can also provide some degree of protection
+against a malicious hypervisor or compromised cloud operator, although
+it should probably not be relied upon as the sole line of defense.
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: leave out the "although it should ...". The upstream spec is not there to explain the quality or lack of quality of the hardware feature. it should focus on how the integration poiints with the openstack software are going to be addressed. 

with other words: We don't have to write long specs to get it approved  ;-)

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+Use Cases
+---------
+
+#. Booting an instance with a flavor containing an SEV-specific
+   `required trait extra spec
+   <https://docs.openstack.org/nova/latest/user/flavors.html#extra-specs-required-traits>`_
+   will cause that instance to be defined and booted on a compute host
+   which provides that trait.  Additionally the instance should be
+   created with additional XML in the libvirt domain definition which
+   causes QEMU to boot the instance with SEV enabled.
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst:48]: as Bryan says, the wording could explain that the traits would trigger two changes in behavior:

* scheduling of the instance on nodes that provide the hardware capability
* defining launch security in the libvirt xml to encrypt the memory of the instance on start

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+        <reducedPhysBits>1</reducedPhysBits>
+        <policy>0x0001</policy>
+      </launchSecurity>
+      ...
+    </domain>
+
+For reference, `the AMDSEV GitHub repository
+<https://github.com/AMDESE/AMDSEV/>`_ provides `a complete example of
+SEV-enabled domain XML
+<https://github.com/AMDESE/AMDSEV/blob/master/xmls/sample.xml>`_.
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: maybe just link the libvirt documentation? https://libvirt.org/formatdomain.html#sev ?

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+
+This trait would also indicate that the libvirt driver should include
+extra XML in the guest's domain definition, to enable SEV security,
+ensure that the VM uses machine type ``q35`` and UEFI, and that
+``iommu_platform=on`` is used for all virtio devices, e.g.::
+
+    <domain type='kvm'>
+      <os>
+        <type arch='x86_64' machine='pc-q35-2.11'>hvm</type>
+        <loader readonly='yes' type='pflash'>/usr/share/qemu/ovmf-x86_64-ms-4m-code.bin</loader>
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst:74]: thats an important part: does SEV require OVMF? if so, this needs to be spelled out.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+
+This work could also mean that images with the property
+``trait:HW_CPU_x86_SEV=required`` would similarly affect the process
+of launching instances.
+
+Immediate follow-up work
+------------------------
+
+It is suggested that follow-up specs be submitted detailing
+the following subsequent work:
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: put it in a separate spec then? we need to clarify what is expected to be implemented in "Stein". Is this part of Stein or not? if so then we need to include it here and not as followup work. if we don't then we don't.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+
+- Support a built-in trait which would be automatically detected per
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst:107]: IMHO this will make the spec approval more complicated because it might not be clear if that is now agreed upon to be implemented or not. if its an optional followup then don't include it in the spec but put it in a separate spec (A new commit in the gerrit sense)

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+          <features>
+            ...
+            <sev supported='yes'/>
+              ...
+            </sev>
+          </features>
+        </domainCapabilities>
+
+  - ``/sys/module/kvm_amd/parameters/sev`` should have the value one
+    to indicate that the kernel has SEV capabilities enabled.
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst:134]: is this needed on top of the libvirt capabilities check?

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+
+- Since `migration of instances between unencrypted and SEV-encrypted
+  states is not supported in either direction
+  <https://github.com/qemu/qemu/commit/8fa4466d77b44f4f58f3836601f31ca5e401485d>`_,
+  prevent nova from resizing an instance from a non-SEV-enabled flavor
+  to an SEV-enabled one, or vice-versa.  Alternatively, nova could
+  catch the error raised by QEMU, which would be propagated via
+  libvirt, and handle it appropriately.
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst:142]: I guess error handling would be preferred and exporting that to the user.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+
+  Similarly, attempts to suspend / resume a domain should either be
+  prevented, or the error caught and handled.
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst:145]: I guess we need to opt for one solution (depending on review feedback)

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 17]
```diff
@@ -0,0 +1,367 @@
...
+
+Alternatives
+------------
+
+#. The idea of initially using a custom trait could be skipped in
+   favour of immediately implementing the built-in trait
+   (``HW_CPU_x86_SEV`` or similar).  This would increase the amount of
+   effort required in order to reach a working prototype, but
+   conversely has the advantage of not requiring the merging of
+   unfinished proof-of-concept code into nova's ``master`` branch.
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: what do you mean by that? upstream will likely not accept "unfinished proof of concept code". the implementation we're proposing under the spec should be complete with unit test coverage

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 18]
```diff
@@ -0,0 +1,367 @@
...
+
+#. Rather than using a single trait to both facilitate the matching
+   of instances requiring SEV with SEV-capable compute hosts *and*
+   indicate to nova's libvirt driver that SEV should be used when
+   booting ...
+
+   However this would create extra friction for the operators defining SEV-enabled
+   flavors,  it is hard to imagine
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: unfinished sentence? leave it out?

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 19]
```diff
@@ -0,0 +1,367 @@
...
+Data model impact
+-----------------
+
+Changes which require modifications to the data model often have a wider impact
+on the system.  The community often has strong opinions on how the data model
+should be evolved, from both a functional and performance perspective. It is
+therefore important to capture and gain agreement as early as possible on any
+proposed changes to the data model.
+
+Questions which need to be addressed by this section include:
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: remove template text and replace it with the trait details from above?

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+            <sev supported='yes'/>
+              ...
+            </sev>
+          </features>
+        </domainCapabilities>
+
+  - ``/sys/module/kvm_amd/parameters/sev`` should have the value one
+    to indicate that the kernel has SEV capabilities enabled.
+
+- Since `migration of instances between unencrypted and SEV-encrypted
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst:136]: does it allow to start unencrypted instances on sev enabled hypervisors? aka is this optional or then a required feature, with other words does the hypervisor become dedicated to sEV workload with this change or not?

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 21]
```diff
@@ -0,0 +1,367 @@
...
+---------------
+
+None, although future work may require extending the REST API so that
+users can verify the hardware's attestation that the memory was
+encrypted correctly by the firmware.
+
+Security impact
+---------------
+
+There are no security impacts for the cloud deployment or cloud
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: for a feature that is all about improved security, this is hard to imagine. 

We need to explain the impact of the launch security token handling. my understanding is that the launch security token is not sensitive information but we might want to consider omiitting it from accidental logging. However, if it is sensitive then this needs to be captured here.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 22]
```diff
@@ -0,0 +1,367 @@
...
+operator.  Tenants of the cloud will benefit from the ability to
+launch encrypted VMs.
+
+It is strongly recommended that `the OpenStack Security Group
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst:204]: boiler plate text? leave it out? either explain what the security group needs to be involved with  or why they do not (because for example that we have excellent security architects and teams on the implementor side already taking care of all aspects).

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 23]
```diff
@@ -0,0 +1,367 @@
...
+<openstack-security@lists.openstack.org>`_ is kept in the loop and
+invited to review each stage of work.
+
+Notifications impact
+--------------------
+
+None
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: we might want to expose theinformation that the instance is running encrypted? a billing cloud provider might want to bill encrypted instances differently than unencrypted ones ("security surcharge")

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 24]
```diff
@@ -0,0 +1,367 @@
...
+
+The end user will harness SEV through the existing mechanisms of
+traits in flavor extra specs and properties.  Later on it may make
+sense to add support for scheduler hints (see the `Future Work`_
+section below).
+
+Performance Impact
+------------------
+
+Encrypted VMs will inevitably perform worse than unencrypted ones
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: unnecessarily pessimistic and irrelevant. The Encryption is having a hardware specific overhead, but thats irrelevant for the spec. the *impact* on *nova* needs to be described here and I think that can be neglected.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 25]
```diff
@@ -0,0 +1,367 @@
...
+due to the encryption overhead.
+
+If compression is being used on swap disks then more storage may be
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst:239]: there is compressed swapping support available? this brings up a good point though: I would believe that memory deduplication efforts like e.g. KSM (kernel samepage merging) would not be effective anymore. Also hugepages I think are affected.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 26]
```diff
@@ -0,0 +1,367 @@
...
+required because the memory of encrypted VMs will not compress to a
+smaller size.
+
+Other deployer impact
+---------------------
+
+The operator will need to define SEV-enabled flavors as described above
+in order for users to be able to use SEV.
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: they als need to deploy a SEV compatible hypervisor if they want to use the feature.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 27]
```diff
@@ -0,0 +1,367 @@
...
+=======
+
+Still TODO
+
+Please discuss the important scenarios needed to test here, as well as
+specific edge cases we should be ensuring work correctly. For each
+scenario please specify if this requires specialized hardware, a full
+openstack environment, or can be simulated inside the Nova tree.
+
+Please discuss how the change will be tested. We especially want to know what
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: remove boiler plate text. please explain why we consider unit test sufficient here.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 28]
```diff
@@ -0,0 +1,367 @@
...
+tempest tests will be added. It is assumed that unit test coverage will be
+added so that doesn't need to be mentioned explicitly, but discussion of why
+you think unit tests are sufficient and we don't need to add more tempest
+tests would need to be included.
+
+Is this untestable in gate given current limitations (specific hardware /
+software configurations available)? If so, are there mitigation plans (3rd
+party testing, gate enhancements, etc).
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: this needs to be filled out from the PTG meeting results.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 29]
```diff
@@ -0,0 +1,367 @@
...
+cloud operators, and the End User Guide would need to be updated if the change
+offers a new feature available through the CLI or dashboard. If a config option
+changes or is deprecated, note here that the documentation needs to be updated
+to reflect this specification's change.
+
+
+Future work
+===========
+
+Looking further ahead, there is scope for several strands of additional
```
Comment (dirkmueller) [inline specs/stein/approved/amd-sev-libvirt-support.rst]: explain that htis is outside scope of Stein? we need to give the "scope" of the work for the next cycle so that upstream can understand when the feature is "done" and when all features are merged, which is important for the feature freeze milestone in the roadmap. as such we need to be very explicit on when we consider the "stein implementation level" complete.

**Review by dirkmueller — CHANGES_REQUESTED** [review]
still a lot of boilerplate text inside, but a good start. imho specs should be opting for more concise wording.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 6 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Not sure I entirely agree but OK, changed in next commit.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 1 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Fixed in next commit.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 7 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: OK, I see what you are saying. Some similar discussion [here](https://www.reddit.com/r/Amd/comments/7of4d2/eli5_is_amds_secure_encrypted_virtualization_sev/).  Reworded in next commit.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 8 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: OTOH some nova developers may have never heard of SEV, and might be inclined to automatic -1 if they can't understand the motivation behind this work.  Context is important.  Reworded in next commit, however.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 2 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Done.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 9 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst:48]: Isn't that exactly what it already says?

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 11 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst:74]: UEFI is explicitly mentioned above.  Is there another way to achieve this except for with OVMF?  But OK, I'll s/UEFI/UEFI via OVMF/

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 10 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Added link.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 12 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Actually this is a deliberate manoeuvre on my part - I put it in to preempt any objections which might come on the basis of having merged code which relies on a hardcoded custom trait name.  This is technical debt, and I expect upstream are less likely to worry about that if they can see that the immediate follow-up work to address it has already been carefully thought through and mapped out.  If they say "actually we want all this work to be done in the same spec" then we simply merge this section with the proposed work above.  If they say "that's fine, just move that to a later spec" then again it's easy.  But if I don't include it then it's likely they will say "this introduces technical debt and we don't see any plan to clean up the mess after".

Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Ditto for the technical debt regarding lack of handling of suspend / resume / migrate etc.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 13 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst:107]: OK, you mean submit two commits at once where the second is on top of the first?  That's actually a pretty good idea, but writing a whole second spec is going to delay submission of this even further (I was hoping to submit it today).  How about I just submit it like this for now, and then suggest to upstream that I can push new patch sets (and potentially new commits) which split out the future work into a separate spec?

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 14 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst:134]: Yes, as explained in great detail on the amd-sev list; that's why I put "all" in italics.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 20 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst:136]: It's optional.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 15 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst:142]: Yeah I would also have to guess on this point, which is why I left it as an implementation detail open for discussion.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 16 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst:145]: Right - either we'll get guidance during spec review, or during (code) commit review, and I think either would be fine.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 17 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: OK, "unfinished" is too alarmist a word - changing that.  It's nothing to do with test coverage; it's about whether it's OK for a reference to a custom trait to be hardcoded in `master`, even for a short time.  Only upstream can answer that.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 18 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: No, there's an important point here so I'll finish it.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 21 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: @bryanstephenson I copied this sentence from the etherpad - can you respond to this please?

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 22 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst:204]: Not boiler-plate.

Well sure, I can explicitly explain that they should be involved because they are (presumably) experts on security within the context of OpenStack with a vested interest in keeping OpenStack secure, and therefore should have the opportunity to review security-related development.  But isn't that stating the obvious?  Indeed, it seems [they are even currently working on AMD CPU capabilities related to security](http://lists.openstack.org/pipermail/openstack-security/2018-October/005724.html).  

Personally, if I was in the OpenStack Security Group and didn't get cc'd on this work, I'd be seriously ****ed off.  It's basic courtesy at very least, and I'm not sure what would be the point of the group if it didn't get to review security-specific development.

Anyway, reworded to make this more explicit.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 23 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Good point - out of scope for this spec, but I'll mention the idea.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 24 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Addressed

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 4 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Done.

BTW totally tangential point - I personally dislike divorcing content from their corresponding links, as it forces the reader to scroll to the bottom to follow the links.  Reference sections made sense in the paper world, but now we have hyperlinks I think we can just use them ;-)  I know Wikipedia doesn't entirely agree with me, but there are probably extra reasons why they do it like that, and at least they use hyperlinked footnotes so that it's two clicks rather than big-scroll-then-click ...

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 25 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst:239]: [**@dirkmueller**](https://github.com/dirkmueller) commented on [11 Oct 2018, 08:23 BST](https://github.com/aspiers/nova-specs/pull/1/files#r224340581 "2018-10-11T07:23:57Z - Replied by Github Reply Comments"):
> there is compressed swapping support available?

I'm not sure.  Isn't that available in SLES, e.g. [zswap](https://www.suse.com/releasenotes/x86_64/SUSE-SLES/12-SP2/#fate-318829) or similar?

> this brings up a good point though: I would believe that memory deduplication efforts like e.g. KSM (kernel samepage merging) would not be effective anymore. Also hugepages I think are affected.

Yep, added thanks.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 26 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Haha yeah OK, I guess it's worth stating the obvious here :-)

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 5 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Sounds good, can you @-mention her GitHub username here?  I can't find it.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 28 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: The email from JP didn't go into much detail about testing but I've put my best guesses in...

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 27 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: I don't think they are sufficient; we probably need functional tests too.  Have added some text about this; please review after push.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 29 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Yes thanks good point, addressed.  This also applies to the "immediate follow-up work" above regarding the built-in trait.  I will try to touch base with Melanie to get advice on this.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 30]
```diff
@@ -0,0 +1,367 @@
...
+
+For reference, `the AMDSEV GitHub repository
+<https://github.com/AMDESE/AMDSEV/>`_ provides `a complete example of
+SEV-enabled domain XML
+<https://github.com/AMDESE/AMDSEV/blob/master/xmls/sample.xml>`_.
+
+This additional functionality could be implemented within
+``nova.virt.libvirt.LibvirtDriver._set_features()``, e.g. using a new
+``nova.virt.libvirt.LibvirtConfigGuestFeatureSEV`` class.
+
```
Comment (evrardjp) [inline specs/stein/approved/amd-sev-libvirt-support.rst:95]: As SEV requires a libvirt >= x, the necessary functions can be implemented with the help of the usual has_min_version decorator to be non-intrusive.

File: specs/stein/approved/amd-sev-libvirt-support.rst [hunk 31]
```diff
@@ -0,0 +1,367 @@
...
+This work could also mean that images with the property
+``trait:HW_CPU_x86_SEV=required`` would similarly affect the process
+of launching instances.
+
```
Comment (evrardjp) [inline specs/stein/approved/amd-sev-libvirt-support.rst:99]: Functional testing will be handled by the libvirt fakedriver, by ensuring the generated xml domain configuration is correct.

**Review by evrardjp — CHANGES_REQUESTED** [review]
A few extra bits I'd consider adding.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 3 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Done, thanks.  Added some others too.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 30 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst:95]: I don't know much about that decorator; how does it work?  Would it support a cloud with a mix of SEV and non-SEV compute hosts?  Does this bring any extra benefit over just using libvirt's virConnectGetDomainCapabilities() API call as described in this spec?  I think I'd prefer to trust libvirt's own opinion of whether it can support SEV, rather than reimplementing that test.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 31 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst:99]: Yeah, let's discuss in Testing section below.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 5 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: Ah, found it.  @jennywei8 - requesting your review.

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 13 above)
Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst:107]: Another option is submitting two specs in a single commit ...

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 30 above)
Comment (evrardjp) [reply specs/stein/approved/amd-sev-libvirt-support.rst:95]: My concern was not about libvirt itself, but for the tests that will be make use of fake driver

File: specs/stein/approved/amd-sev-libvirt-support.rst (hunk 21 above)
Comment (bryanstephenson) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: I will assume that "launch security token" here means the indication that an encrypted VM is needed. If you mean something else please correct this assumption and provide a reference to the "launch security token" information. I do not see security impacts of adding the ability to encrypt a VM given the simple way we are enabling the feature. I do not view the information that a VM is encrypted or not to be sensitive information, but some customers might feel that way and want that choice kept out of log files. I recommend using this text instead of the current text:

This change does not add or handle any secret information. The choice to encrypt a VM or not may be considered somewhat sensitive information in some environments, but this information will be logged to facilitate troubleshooting in cases where the SEV functionality is suspected of causing an issue. If user feedback indicates that this information should not be logged then this decision could be changed in future.

Comment (aspiers) [reply specs/stein/approved/amd-sev-libvirt-support.rst]: [**@bryanstephenson**](https://github.com/bryanstephenson) commented on [11 Oct 2018, 17:07 BST](https://github.com/aspiers/nova-specs/pull/1/files#r224509788 "2018-10-11T16:07:06Z - Replied by Github Reply Comments"):
> I will assume that "launch security token" here means the indication that an encrypted VM is needed.

ICBW but I don't think that's what he meant - normally "token" refers to a secret or credential of some sort.

> If you mean something else please correct this assumption and provide a reference to the "launch security token" information.

By "token" I guess he was assuming there is some secret token which needs to be passed around as part of the interactions between AMD hardware and the user, but I don't think this is the case.

Anyway, I've updated the text based on your wording - thanks!

**Comment by aspiers** [comment]
This is now upstream - please continue review process there:

https://review.openstack.org/#/c/609779/
